runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -15665,10 +15665,60 @@ J9::Z::TreeEvaluator::BCDCHKEvaluatorImpl(TR::Node * node,
                                           bool isUseVector)
    {
    TR_Debug* debugObj = cg->getDebug();
-   TR::Node * secondChild = node->getChild(1);
+   TR::Node* secondChild = node->getChild(1);
+   TR::Compilation* comp = cg->comp();
+   TR::Register* bcdOpResultReg = NULL;
+
    bool isVariableParam = pdopNode->getOpCodeValue() == TR::lcall ||
                           pdopNode->getOpCodeValue() == TR::icall;
 
+   // Skip the evaluation of pdshlOverflow if its child i2pd (or l2pd) is already evaluated.
+   //
+   // If vector instructions are used, the pdshlOverflow+i2pd structure only requires one VCVD(G) instruction to achieve
+   // the desired functionality. Effectively, the pdshlOverflow does not need any evaluation. The value (register) associated with
+   // i2pd can be copied over to the pdshlOverflow node. This also implies that the
+   // pdshlOverflow node is not an exceptioning node. Thus, generating an OOL sequence for a non-exceptioning node no only
+   // violates the definition of BCDCHK node, but also can make RA generate bad GC map and eventually cause GC error during runtime
+   //
+   if (isUseVector)
+      {
+      bool hasEvaluatedI2PD = pdopNode->getOpCodeValue() == TR::pdshlOverflow
+              && (pdopNode->getFirstChild()->getOpCodeValue() == TR::i2pd ||
+                  pdopNode->getFirstChild()->getOpCodeValue() == TR::l2pd)
+              && pdopNode->getFirstChild()->getRegister() != NULL;
+
+      if (hasEvaluatedI2PD && pdopNode->getRegister() == NULL)
+         {
+         cg->evaluate(pdopNode);
+         bcdOpResultReg = pdopNode->getRegister();
+         cg->decReferenceCount(pdopNode);
+         }
+
+      if (bcdOpResultReg != NULL || hasEvaluatedI2PD)
```

Comment:
This seems incorrect. If `bcdOpResultReg != NULL` then you must have entered the above `if`, which means `hasEvaluatedI2PD == true` and `pdopNode->getRegister() == NULL`. This implies you must have generated a `VSRP` instruction which can overflow so a BCDCHK is needed here.

Reply:
I should've left more comment for this to clarify more. There are some implications if the JIT reaches this line. 


If `hasEvaluatedI2PD=true`, it implies that the `pdshlOverflow` is NOT evaluated and it's child `i2pd` is evaluated "previously in another piece of mainline for another BCDCHK node".


      if (hasEvaluatedI2PD && pdopNode->getRegister() == NULL)
         {
         cg->evaluate(pdopNode);
         bcdOpResultReg = pdopNode->getRegister();
         cg->decReferenceCount(pdopNode);
         }
Also, evaluating the pdopNode here simply copies the `i2pd` node's VRF register to the `pdshlOverflow` node. This does not generate a `VSRP`... A `VSRP` is not necessary as the vector `VCVD` does both the conversion and truncation. This is where vector BCD differ from the SS instructions.

Reply:
> Also, evaluating the pdopNode here simply copies the i2pd node's VRF register to the pdshlOverflow node. This does not generate a VSRP... A VSRP is not necessary as the vector VCVD does both the conversion and truncation. This is where vector BCD differ from the SS instructions.

This is incorrect then, since the decimal precision of the `i2pd` node may not match that of the `pdshlOverflow` node, in which case we must generate a `VSRP` for the truncation. See my comment in the above review snippet for the two scenarios. We need to address this.